cranelift/codegen/src/machinst/lower.rs

```diff
@@ -0,0 +1,727 @@
+//! This module implements lowering (instruction selection) from Cranelift IR
+//! to machine instructions with virtual registers. This is *almost* the final
+//! machine code, except for register allocation.
+
+use crate::binemit::CodeSink;
+use crate::entity::SecondaryMap;
+use crate::inst_predicates::has_side_effect;
+use crate::ir::instructions::BranchInfo;
+use crate::ir::{
+    Block, ExternalName, Function, GlobalValueData, Inst, InstructionData, MemFlags, Opcode,
+    Signature, SourceLoc, Type, Value, ValueDef,
+};
+use crate::isa::registers::RegUnit;
+use crate::machinst::{
+    ABIBody, BlockIndex, MachInst, MachInstEmit, VCode, VCodeBuilder, VCodeInst,
+};
+use crate::num_uses::NumUses;
+
+use regalloc::Function as RegallocFunction;
+use regalloc::{RealReg, Reg, RegClass, Set, VirtualReg, Writable};
+
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use log::debug;
+use smallvec::SmallVec;
+use std::collections::VecDeque;
+use std::ops::Range;
+
+/// A context that machine-specific lowering code can use to emit lowered instructions. This is the
+/// view of the machine-independent per-function lowering context that is seen by the machine
+/// backend.
+pub trait LowerCtx {
```

Comment:
The intention is to also eventually have other `Lower` impls. For one, there has been some conversation about verification efforts that plug in a different impl to get symbolic lowerings out (this is very handwavy now, but we want to leave the door open). For another, we might eventually want to drive the lowering logic with a different IR, and so we're trying to keep as much behind this API as possible (it's not perfect, e.g. `Opcode` type, but this will help with that goal). Everything is monomorphized, so the only cost now should be a little extra type verbosity.